### PR TITLE
feat: neutron dao with interchain security (NTRN-206)

### PR DESCRIPTION
**Pull request by quasisamurai**
[Task 1](https://p2pvalidator.atlassian.net/browse/NTRN-169)
[Task 2](https://p2pvalidator.atlassian.net/browse/NTRN-206)

This PR implements a neutron's vision of governance. It uses [admin module](https://github.com/cosmos/composer) in bound with smart contract bindings, provided w this pr.

testing: 
- `make install`
- `make init`
- from [this pr](https://github.com/neutron-org/neutron-dao/pull/8) `./test_proposal.sh`

### CODE REVIEW

File: wasmbinding/message_plugin.go [hunk 1]
```diff
@@ -79,6 +77,9 @@ func (m *CustomMessenger) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddre
 			return m.ibcTransfer(ctx, contractAddr, *contractMsg.IBCTransfer)
 		}
 		if contractMsg.SubmitProposal != nil {
+			ctx.Logger().Debug("PRAPASAL",
```
Comment (zavgorodnii) [inline wasmbinding/message_plugin.go]: bruh lol

Comment (quasisamurai) [reply wasmbinding/message_plugin.go]: D:

File: go.mod [hunk 2]
```diff
@@ -130,7 +130,7 @@ require (
 )
 
 replace (
-	github.com/CosmWasm/wasmd v0.28.0 => github.com/neutron-org/wasmd v0.28.0-keep-ibc-events
+	github.com/CosmWasm/wasmd v0.28.0 => github.com/neutron-org/wasmd v0.28.1-0.20221130133614-9661821bf976
```
Comment (oldremez) [inline go.mod:133]: We should keep in mind to replace it with `neutron` branch once PR to wasmd will be merged

**Comment by zavgorodnii** [comment]
also i don't see parameter change proposals in the testing script, please add it

File: network/init.sh [hunk 3]
```diff
@@ -65,19 +68,47 @@ $BINARY add-genesis-account $($BINARY --home $CHAIN_DIR/$CHAINID_1 keys show dem
...
 # Instantiate the contract
-INIT_CONTRACT="$(printf '{"owner":"%s"}' "${ADMIN_ADDRESS}")"
+#INIT_CONTRACT="$(printf '{"owner":"%s"}' "${ADMIN_ADDRESS}")"
+INIT="$(printf '{"denom":"stake"}')"
+#TODO: fix quotes issue
+DAO_INIT="'""'"
 #echo "Instantiate"
-$BINARY add-wasm-message  instantiate-contract 1 "$INIT_CONTRACT" --run-as neutron1mjk79fjjgpplak5wq838w0yd982gzkyf8fxu8u --admin ${ADMIN_ADDRESS}  --label "DAO"  --home $CHAIN_DIR/$CHAINID_1
+$BINARY add-wasm-message  instantiate-contract 1 ${INIT} --run-as neutron1m9l358xunhhwds0568za49mzhvuxx9ux8xafx2 --admin ${ADMIN_ADDRESS}  --label "staking"  --home $CHAIN_DIR/$CHAINID_1
+$BINARY add-wasm-message  instantiate-contract 2 '{
```
Comment (sotnikov-s) [inline network/init.sh]: such a strange json formatting. is it possible to shift the whole body to the left closer to the beginning with a small indentation?

Comment (sotnikov-s) [reply network/init.sh]: maybe let's move the whole json to a variable as you've done with INIT?

File: network/init.sh [hunk 4]
```diff
@@ -65,19 +68,47 @@ $BINARY add-genesis-account $($BINARY --home $CHAIN_DIR/$CHAINID_1 keys show dem
...
+$BINARY add-wasm-message store ${STAKING_CONTRACT} --output json --run-as neutron1m9l358xunhhwds0568za49mzhvuxx9ux8xafx2 --keyring-backend=test --home $CHAIN_DIR/$CHAINID_1
+$BINARY add-wasm-message store ${DAO_CONTRACT} --output json  --run-as neutron1m9l358xunhhwds0568za49mzhvuxx9ux8xafx2 --home $CHAIN_DIR/$CHAINID_1
+$BINARY add-wasm-message store ${PROPOSAL_CONTRACT} --output json  --run-as neutron1m9l358xunhhwds0568za49mzhvuxx9ux8xafx2 --home $CHAIN_DIR/$CHAINID_1
+$BINARY add-wasm-message store ${VOTING_CONTRACT} --output json  --run-as neutron1m9l358xunhhwds0568za49mzhvuxx9ux8xafx2 --home $CHAIN_DIR/$CHAINID_1
 # Instantiate the contract
-INIT_CONTRACT="$(printf '{"owner":"%s"}' "${ADMIN_ADDRESS}")"
+#INIT_CONTRACT="$(printf '{"owner":"%s"}' "${ADMIN_ADDRESS}")"
+INIT="$(printf '{"denom":"stake"}')"
+#TODO: fix quotes issue
+DAO_INIT="'""'"
```
Comment (sotnikov-s) [inline network/init.sh]: unused variable

File: network/init.sh [hunk 5]
```diff
@@ -65,19 +68,47 @@ $BINARY add-genesis-account $($BINARY --home $CHAIN_DIR/$CHAINID_1 keys show dem
...
-echo "Creating and collecting gentx..."
-$BINARY gentx val1 7000000000stake --home $CHAIN_DIR/$CHAINID_1 --chain-id $CHAINID_1 --keyring-backend test
-$BINARY gentx val2 7000000000stake --home $CHAIN_DIR/$CHAINID_2 --chain-id $CHAINID_2 --keyring-backend test
-$BINARY collect-gentxs --home $CHAIN_DIR/$CHAINID_1
-$BINARY collect-gentxs --home $CHAIN_DIR/$CHAINID_2
-
 echo "Initializing dao contract in genesis..."
 # Upload the dao contract
-$BINARY add-wasm-message store ${DAO_CONTRACT} --output json  --run-as neutron1mjk79fjjgpplak5wq838w0yd982gzkyf8fxu8u --home $CHAIN_DIR/$CHAINID_1
+$BINARY add-wasm-message store ${STAKING_CONTRACT} --output json --run-as neutron1m9l358xunhhwds0568za49mzhvuxx9ux8xafx2 --keyring-backend=test --home $CHAIN_DIR/$CHAINID_1
```
Comment (sotnikov-s) [inline network/init.sh]: let's move `neutron1m9l358xunhhwds0568za49mzhvuxx9ux8xafx2` to a variable?

File: network/init.sh (hunk 3 above)
Comment (quasisamurai) [reply network/init.sh]: i already talked about this section in tg,  nvm

**Comment by NeverHappened** [comment]
Tried it, script works.

**Review by albertandrejev — DISMISSED** [review]

**Review by zavgorodnii — DISMISSED** [review]

File: network/init.sh [hunk 6]
```diff
@@ -1,12 +1,16 @@
 #!/bin/bash
 
-ADMIN_ADDRESS=neutron14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s5c2epq
-NEUTROND_BINARY=neutrond
+
+ADMIN_ADDRESS=neutron1m9l358xunhhwds0568za49mzhvuxx9ux8xafx2
+BINARY=neutrond
```
Comment (albertandrejev) [inline network/init.sh]: Please rename this variable to `NEUTROND_BINARY` please.

**Review by albertandrejev — APPROVED** [review]

**Review by zavgorodnii — APPROVED** [review]

**Pull request merged**